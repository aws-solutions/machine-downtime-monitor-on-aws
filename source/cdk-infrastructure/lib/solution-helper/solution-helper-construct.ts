// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct, CustomResource, Duration, CfnCondition, CfnCustomResource, Aws } from '@aws-cdk/core';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import { Table } from '@aws-cdk/aws-dynamodb';
import { Function as LambdaFunction, Runtime, Code, CfnFunction } from '@aws-cdk/aws-lambda';

export interface ISolutionHelperProps {
    readonly sourceCodeBucketName: string;
    readonly sourceCodeKeyPrefix: string;
    readonly sendAnonymousData: string;
    readonly anonymousUsageCondition: CfnCondition;
    readonly solutionVersion: string;
    readonly solutionId: string;
    readonly deployM2CParameter: string;
}

export class SolutionHelper extends Construct {
    private readonly sourceCodeBucket: IBucket;
    public readonly customResourceLambda: LambdaFunction;
    private readonly sourceCodeKeyPrefix: string;
    public readonly anonymousDataUUID: string;
    public readonly lowerCaseStackName: string;

    constructor(scope: Construct, id: string, props: ISolutionHelperProps) {
        super(scope, id);

        this.sourceCodeBucket = Bucket.fromBucketName(this, 'sourceCodeBucket', props.sourceCodeBucketName);
        this.sourceCodeKeyPrefix = props.sourceCodeKeyPrefix;

        const generateSolutionConstantsLambda = new LambdaFunction(this, 'GenerateSolutionConstantsLambda', {
            runtime: Runtime.NODEJS_14_X,
            handler: 'solution-helper/index.handler',
            timeout: Duration.seconds(3),
            description: 'Generates an anonymous UUID and a lower case stack name for sending metrics and resource creation',
            code: Code.fromBucket(this.sourceCodeBucket, [props.sourceCodeKeyPrefix, 'solution-helper.zip'].join('/')),
            environment: { STACK_NAME: Aws.STACK_NAME }
        });

        // Suppress CFN Nag warnings
        (generateSolutionConstantsLambda.node.findChild('Resource') as CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    id: 'W58',
                    reason: 'CloudWatch permissions are granted by the LambdaBasicExecutionRole'
                },
                {
                    id: 'W89',
                    reason: 'VPC for Lambda is not needed. This serverless architecture does not deploy a VPC.'
                },
                {
                    id: 'W92',
                    reason: 'ReservedConcurrentExecutions is not needed for this Lambda function.'
                }
            ]
        });

        const generateSolutionConstantsCustomResource = new CustomResource(this, 'GenerateSolutionConstantsCustomResource', {
            serviceToken: generateSolutionConstantsLambda.functionArn,
            properties: { Action: 'GENERATE_SOLUTION_CONSTANTS' }
        });
        this.anonymousDataUUID = generateSolutionConstantsCustomResource.getAttString('AnonymousDataUUID');
        this.lowerCaseStackName = generateSolutionConstantsCustomResource.getAttString('LowerCaseStackName');

        this.customResourceLambda = new LambdaFunction(this, 'CustomResourceLambda', {
            runtime: Runtime.NODEJS_14_X,
            handler: 'solution-helper/index.handler',
            timeout: Duration.minutes(2),
            description: 'Performs various Solution lifecycle actions when invoked as a CloudFormation Custom Resource',
            code: Code.fromBucket(this.sourceCodeBucket, [props.sourceCodeKeyPrefix, 'solution-helper.zip'].join('/')),
            environment: {
                SEND_ANONYMOUS_DATA: props.sendAnonymousData,
                ANONYMOUS_DATA_UUID: this.anonymousDataUUID,
                SOLUTION_ID: props.solutionId,
                SOLUTION_VERSION: props.solutionVersion
            }
        });

        // Suppress CFN Nag warnings
        (this.customResourceLambda.node.findChild('Resource') as CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    id: 'W58',
                    reason: 'CloudWatch permissions are granted by the LambdaBasicExecutionRole'
                },
                {
                    id: 'W89',
                    reason: 'VPC for Lambda is not needed. This serverless architecture does not deploy a VPC.'
                },
                {
                    id: 'W92',
                    reason: 'ReservedConcurrentExecutions is not needed for this Lambda function.'
                }
            ]
        });

        const solutionLifecycleMetricCustomResource = new CfnCustomResource(this, 'SolutionLifecycleMetricCustomResource', { serviceToken: this.customResourceLambda.functionArn });
        solutionLifecycleMetricCustomResource.addPropertyOverride('Action', 'SOLUTION_LIFECYCLE');
        solutionLifecycleMetricCustomResource.addPropertyOverride('SolutionParameters', { DeployM2C: props.deployM2CParameter });
        solutionLifecycleMetricCustomResource.cfnOptions.condition = props.anonymousUsageCondition;
    }

    public setupCopyAssetsCustomResource(props: ISetupCopyAssetsCustomResourceProps) {
        // Allows the custom resource to read the static assets for the Amplify front-end from the source code bucket
        this.sourceCodeBucket.grantRead(this.customResourceLambda, `${this.sourceCodeKeyPrefix}/*`);

        // Allows the custom resource to place the static assets for the Amplify front-end into the hosting bucket
        props.hostingBucket.grantPut(this.customResourceLambda);

        new CustomResource(this, 'CopyAssetsCustomResource', {  // NOSONAR: typescript:S1848
            serviceToken: this.customResourceLambda.functionArn,
            properties: {
                Action: 'CONFIGURE_UI',
                DestinationBucket: props.hostingBucket.bucketName,
                SrcBucket: this.sourceCodeBucket.bucketName,
                SrcPath: this.sourceCodeKeyPrefix,
                WebUIManifestFileName: 'web-ui-manifest.json',
                WebUIStaticFileNamePrefix: 'web-ui/',
                WebUIConfigFileName: 'web-ui-config.js',
                IdentityPoolId: props.identityPoolId,
                UserPoolId: props.userPoolId,
                UserPoolClientId: props.userPoolClientId,
                ApiEndpoint: props.graphQLEndpoint
            }
        });
    }

    public setupConfigureMachineDataCustomResource(props: ISetupConfigureMachineDataCustomResourceProps) {
        // Allows the custom resource to put the default configuration into the tables
        props.configTable.grantWriteData(this.customResourceLambda);
        props.uiReferenceTable.grantWriteData(this.customResourceLambda);

        new CustomResource(this, 'ConfigureMachineDataCustomResource', {    // NOSONAR: typescript:S1848
            serviceToken: this.customResourceLambda.functionArn,
            properties: {
                Action: 'CONFIGURE_MACHINE_DATA',
                ConfigId: 'DEFAULT',
                ConfigTableName: props.configTable.tableName,
                UIReferenceTableName: props.uiReferenceTable.tableName,
                MessageFormat: {
                    msgFormatDataAliasDelimiter: '/',
                    msgFormatDataMessageAliasKeyName: 'name',
                    msgFormatDataMessageQualityKeyName: 'quality',
                    msgFormatDataMessagesKeyName: 'messages',
                    msgFormatDataMessageTimestampFormat: 'YYYY-MM-DD HH:mm:ss.SSSSSSZZ',
                    msgFormatDataMessageTimestampKeyName: 'timestamp',
                    msgFormatDataMessageValueKeyName: 'value'
                },
                UIReferenceMapping: {
                    uiReferenceMappingLineKeys: '2',
                    uiReferenceMappingLocationKeys: '0/1'
                }
            }
        });
    }
}

interface ISetupCopyAssetsCustomResourceProps {
    hostingBucket: Bucket;
    identityPoolId: string;
    userPoolId: string;
    userPoolClientId: string;
    graphQLEndpoint: string;
}

interface ISetupConfigureMachineDataCustomResourceProps {
    configTable: Table;
    uiReferenceTable: Table;
}
