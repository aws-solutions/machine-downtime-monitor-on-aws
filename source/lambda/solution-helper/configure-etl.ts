// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IConfigETLRequest, IConfigETLRequestProps, IQuickSightManifest } from './custom-resource-utils';

// AWS Clients
import S3 from 'aws-sdk/clients/s3';
const s3 = new S3();

/**
 * Handles ETL configuration.
 * @param {IConfigETLRequest} event Lambda event
 * @returns {Promise<string>} Custom resource result message
 */
export async function handleEtlConfiguration(event: IConfigETLRequest): Promise<string> {
  switch (event.RequestType) {
    case 'Create':
      await Promise.all([
        copyS3Objects(event.ResourceProperties),
        createMachineInformationFiles(event.ResourceProperties)
      ]);

      break;
    case 'Update':
      await copyS3Objects(event.ResourceProperties);
      break
    default:
      return `No action needed for ${event.RequestType}`;
  }

  return `${event.RequestType} completed OK`;
}

/**
 * Copies S3 objects from the source bucket to the destination bucket.
 * @param {IConfigETLRequestProps} props Custom resource properties
 */
async function copyS3Objects(props: IConfigETLRequestProps) {
  const files = props.GlueJobScripts;

  await Promise.all(files.map(async (file: string) => {
    const params: S3.CopyObjectRequest = {
      Bucket: props.DestinationBucket,
      CopySource: `${props.SourceBucket}/${props.SourcePrefix}/${props.GlueJobScriptsPrefix}/${file}`,
      Key: `${props.GlueJobScriptsPrefix}/${file}`
    };

    console.log('Copying file:', JSON.stringify(params, null, 2));
    const response = await s3.copyObject(params).promise();
    console.log('File copied to s3:', JSON.stringify(response, null, 2));
  }));
}

/**
 * Creates machine information metadata files.
 * These file are used to create QuickSight resources.
 * @param {IConfigETLRequestProps} props Custom resource properties
 */
async function createMachineInformationFiles(props: IConfigETLRequestProps) {
  const destinationBucket = props.DestinationBucket;
  const csvPrefix = props.CsvPrefix;
  const manifestPrefix = props.ManifestPrefix;
  const machineInformationPrefix = props.MachineInformationPrefix;
  const machineConfigInformationPrefix = props.MachineConfigInformationPrefix;

  const machineInformationCsv = `${machineInformationPrefix}.csv`;
  const machineConfigInformationCsv = `${machineConfigInformationPrefix}.csv`;
  const machineInformationManifest = `${machineInformationPrefix}_${manifestPrefix}.json`;
  const machineConfigInformationManifest = `${machineConfigInformationPrefix}_${manifestPrefix}.json`;

  const machineInformationCsvBody = 'id,machine_name,location,line';
  const machineConfigInformationCsvBody = 'id,status_tag,down_value';
  const machineInformationManifestBody = getQuickSightManifest(destinationBucket, csvPrefix, machineInformationCsv);
  const machineConfigInformationManifestBody = getQuickSightManifest(destinationBucket, csvPrefix, machineConfigInformationCsv);

  await Promise.all([
    putObjectToS3Bucket(destinationBucket, `${csvPrefix}/${machineInformationCsv}`, machineInformationCsvBody),
    putObjectToS3Bucket(destinationBucket, `${csvPrefix}/${machineConfigInformationCsv}`, machineConfigInformationCsvBody),
    putObjectToS3Bucket(destinationBucket, `${manifestPrefix}/${machineInformationManifest}`, JSON.stringify(machineInformationManifestBody)),
    putObjectToS3Bucket(destinationBucket, `${manifestPrefix}/${machineConfigInformationManifest}`, JSON.stringify(machineConfigInformationManifestBody))
  ]);
}

/**
 * Puts an object into the S3 bucket with the provided object key.
 * @param {string} bucket S3 bucket to put the object
 * @param {string} key S3 object key to put the object
 * @param {string} body S3 object string body
 */
async function putObjectToS3Bucket(bucket: string, key: string, body: string) {
  const params: S3.PutObjectRequest = { Bucket: bucket, Key: key, Body: body };

  console.log('Putting file:', JSON.stringify(params, null, 2));
  const response = await s3.putObject(params).promise();
  console.log('File put to s3:', JSON.stringify(response, null, 2));
}

/**
 * Gets QuickSight manifest JSON object.
 * @param {string} bucket S3 bucket where the CSV file is
 * @param {string} csvPrefix S3 bucket prefix where the CSV file is
 * @param {string} csvFileName CSV file name in the S3 bucket
 * @returns {IQuickSightManifest} The QuickSight manifest JSON object
 */
function getQuickSightManifest(bucket: string, csvPrefix: string, csvFileName: string): IQuickSightManifest {
  return {
    fileLocations: [
      { URIs: [`s3://${bucket}/${csvPrefix}/${csvFileName}`] }
    ],
    globalUploadSettings: {
      format: 'CSV',
      delimiter: ',',
      textqualifier: '\'',
      containsHeader: 'true'
    }
  }
}