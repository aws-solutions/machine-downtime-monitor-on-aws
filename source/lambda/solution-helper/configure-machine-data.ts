// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IConfigMachineDataRequest } from './custom-resource-utils';
import { ConfigType, IMessageFormatConfigItem, IUIReferenceMappingConfigItem } from '../util/gql-schema-interfaces';
import { getOptions } from '../util/metrics';
import { DocumentClient } from 'aws-sdk/clients/dynamodb';
const ddbDocClient = new DocumentClient(getOptions());

export async function handleConfigureMachinieData(event: IConfigMachineDataRequest): Promise<string> {
    if (event.RequestType === 'Create') {
        if (!event.ResourceProperties.ConfigId) { throw new Error('ConfigId was not supplied'); }

        if (event.ResourceProperties.MessageFormat) {
            await putConfig(
                event.ResourceProperties.ConfigId,
                ConfigType.MESSAGE_FORMAT,
                event.ResourceProperties.ConfigTableName,
                event.ResourceProperties.MessageFormat);
        }

        if (event.ResourceProperties.UIReferenceMapping) {
            await putConfig(
                event.ResourceProperties.ConfigId,
                ConfigType.UI_REFERENCE_MAPPING,
                event.ResourceProperties.UIReferenceTableName,
                event.ResourceProperties.UIReferenceMapping);
        }
        return `${event.RequestType} completed OK`;
    } else {
        return `No action needed for ${event.RequestType}`;
    }
}

async function putConfig(id: string, type: ConfigType, TableName: string, properties: IMessageFormatConfigItem | IUIReferenceMappingConfigItem): Promise<void> {
    const putParams: DocumentClient.PutItemInput = {
        TableName,
        Item: { ...properties, id, type }
    };

    console.log('Putting config', JSON.stringify(putParams, null, 2));
    await ddbDocClient.put(putParams).promise();
    console.log('Done')
}
