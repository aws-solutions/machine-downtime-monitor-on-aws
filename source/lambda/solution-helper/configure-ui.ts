// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IConfigUIRequest, IConfigUIRequestProps } from './custom-resource-utils';
import { getOptions } from '../util/metrics';
import S3 from 'aws-sdk/clients/s3';
const S3Client = new S3(getOptions());

const { AWS_REGION } = process.env;

export async function handleConfigureUI(event: IConfigUIRequest): Promise<string> {
    switch (event.RequestType) {
        case 'Create':
        case 'Update':
            await copyStaticUIAssets(event.ResourceProperties);
            await putAmplifyConfig(event.ResourceProperties);
            return `${event.RequestType} completed OK`;
        default:
            return `No action needed for ${event.RequestType}`;
    }
}

/**
 * Copies the static assets listed in the Web UI manifest file to the Web UI S3 bucket
 * @param props {IConfigUIRequestProps} Custom Resource Properties
 */
async function copyStaticUIAssets(props: IConfigUIRequestProps): Promise<void> {
    // get file manifest from s3
    const getManifestParams: S3.GetObjectRequest = {
        Bucket: props.SrcBucket,
        Key: `${props.SrcPath}/${props.WebUIManifestFileName}`
    };

    console.log('Getting manifest file:', JSON.stringify(getManifestParams, null, 2));
    const data = await S3Client.getObject(getManifestParams).promise();
    const manifest: string[] = JSON.parse(data.Body.toString());
    console.log('Manifest:', JSON.stringify(manifest, null, 2));

    // Loop through manifest and copy files to the destination bucket
    await Promise.all(manifest.map(async (file: string) => {
        let params: S3.CopyObjectRequest = {
            Bucket: props.DestinationBucket,
            CopySource: `${props.SrcBucket}/${props.SrcPath}/${file}`,
            Key: (file.startsWith(props.WebUIStaticFileNamePrefix) && file !== props.WebUIStaticFileNamePrefix) ? file.split(props.WebUIStaticFileNamePrefix).slice(1).join('') : file
        };

        console.log('Copying:', JSON.stringify(params, null, 2));
        let resp = await S3Client.copyObject(params).promise();
        console.log('file copied to s3:', resp);
    }));
}

/**
 * Places a file in the Web UI S3 bucket for use by Amplify when the UI is loaded and configured
 * @param props {IConfigUIRequestProps} Custom Resource Properties
 */
async function putAmplifyConfig(props: IConfigUIRequestProps): Promise<void> {
    const webUIConfig = {
        Auth: {
            mandatorySignIn: true,
            region: AWS_REGION,
            identityPoolId: props.IdentityPoolId,
            userPoolId: props.UserPoolId,
            userPoolWebClientId: props.UserPoolClientId
        },
        "aws_appsync_graphqlEndpoint": props.ApiEndpoint,
        "aws_appsync_region": AWS_REGION,
        "aws_appsync_authenticationType": "AWS_IAM"
    };

    const webUIConfigFile = `const webUIAWSConfig = ${JSON.stringify(webUIConfig, null, 2)};`;

    console.log('Web UI Config Contents: ', webUIConfigFile);

    const putObjectProps: S3.PutObjectRequest = {
        Bucket: props.DestinationBucket,
        Key: props.WebUIConfigFileName,
        Body: Buffer.from(webUIConfigFile),
        ContentType: 'application/javascript'
    };

    console.log(`Putting Web UI Config: ${props.DestinationBucket}/${props.WebUIConfigFileName}`);
    await S3Client.putObject(putObjectProps).promise();
    console.log('Successfully put Web UI Config');
}
