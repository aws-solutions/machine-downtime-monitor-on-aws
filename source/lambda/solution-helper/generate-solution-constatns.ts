// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IGenerateUUIDRequest } from './custom-resource-utils';
import { v4 as uuidV4 } from 'uuid';

interface HandlerOutput {
    anonymousUUID?: string;
    lowerCaseStackName?: string;
}
export function handleGenerateSolutionConstants(event: IGenerateUUIDRequest): HandlerOutput {
    if (event.RequestType === 'Create') {
        return {
            anonymousUUID: uuidV4(),
            lowerCaseStackName: process.env.STACK_NAME.toLowerCase()
        };
    }

    return {};
}
