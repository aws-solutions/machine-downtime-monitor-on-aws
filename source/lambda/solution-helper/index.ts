// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import {
    ICustomResourceRequest, ILambdaContext, ICompletionStatus,
    IConfigUIRequest, IConfigMachineDataRequest,
    IGenerateUUIDRequest, ISolutionLifecycleMetricRequest, IConfigETLRequest,
    ICreateQuickSightRequest, CustomResourceActions, StatusTypes,
} from './custom-resource-utils';
import { handleConfigureUI } from './configure-ui';
import { handleConfigureMachinieData } from './configure-machine-data';
import { handleGenerateSolutionConstants } from './generate-solution-constatns';
import { handleSolutionLifecycleMetric } from './solution-lifecycle-metric';
import { handleEtlConfiguration } from './configure-etl';
import { handleQuickSightResourceCreation } from './create-quicksight';

exports.handler = async (event: ICustomResourceRequest, context: ILambdaContext) => {
    console.log('Received Event', JSON.stringify(event, null, 2));

    const completionStatus: ICompletionStatus = {
        Status: StatusTypes.Success,
        Data: {}
    };

    try {
        if (!event.ResourceProperties.Action) {
            throw new Error('Custom Resource Action was not supplied');
        }

        switch (event.ResourceProperties.Action) {
            case CustomResourceActions.CONFIGURE_UI:
                completionStatus.Data.Message = await handleConfigureUI(event as IConfigUIRequest);
                break;
            case CustomResourceActions.CONFIGURE_MACHINE_DATA:
                completionStatus.Data.Message = await handleConfigureMachinieData(event as IConfigMachineDataRequest);
                break;
            case CustomResourceActions.GENERATE_SOLUTION_CONSTANTS:
                const handlerOutput = handleGenerateSolutionConstants(event as IGenerateUUIDRequest);
                if (!handlerOutput.anonymousUUID) {
                    completionStatus.Data.Message = `No action needed for ${event.ResourceProperties.Action}`;
                } else {
                    completionStatus.Data.AnonymousDataUUID = handlerOutput.anonymousUUID;
                    completionStatus.Data.LowerCaseStackName = handlerOutput.lowerCaseStackName;
                }
                break;
            case CustomResourceActions.SOLUTION_LIFECYCLE:
                completionStatus.Data.Message = await handleSolutionLifecycleMetric(event as ISolutionLifecycleMetricRequest);
                break;
            case CustomResourceActions.CONFIGURE_ETL:
                completionStatus.Data.Message = await handleEtlConfiguration(event as IConfigETLRequest);
                break;
            case CustomResourceActions.CREATE_QUICKSIGHT:
                completionStatus.Data.Message = await handleQuickSightResourceCreation(event as ICreateQuickSightRequest);
                break;
            default:
                throw new Error(`Unknown Custom Resource Action: ${event.ResourceProperties.Action}`);
        }
    } catch (err) {
        console.error(err);
        completionStatus.Data.Error = err.message;
        completionStatus.Status = StatusTypes.Failed;
    }

    const cfnResponse = await respondToCloudFormation(event, context.logStreamName, completionStatus);
    console.log(`CloudFormation Response: ${cfnResponse.statusText} (${cfnResponse.status})`);

    return completionStatus;
}

async function respondToCloudFormation(event: ICustomResourceRequest, logStreamName: string, completionStatus: ICompletionStatus): Promise<AxiosResponse> {
    const responseBody = JSON.stringify({
        Status: completionStatus.Status.toString(),
        Reason: `See the details in CloudWatch Log Stream: ${logStreamName}`,
        PhysicalResourceId: event.LogicalResourceId,
        StackId: event.StackId,
        RequestId: event.RequestId,
        LogicalResourceId: event.LogicalResourceId,
        Data: completionStatus.Data,
    });

    const config: AxiosRequestConfig = {
        headers: {
            'content-type': '',
            'content-length': responseBody.length
        }
    };

    console.log('Responding to CloudFormation', responseBody);
    return await axios.put(event.ResponseURL, responseBody, config);
}
