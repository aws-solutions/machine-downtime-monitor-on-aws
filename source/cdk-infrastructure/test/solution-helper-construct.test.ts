// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { CfnCondition, Stack, Fn } from '@aws-cdk/core';
import { SolutionHelper } from '../lib/solution-helper/solution-helper-construct';

test('Solution helper snapshot and variables', () => {
  // PREPARE
  const stack = new Stack();
  const anonymousDataCondition = new CfnCondition(
    stack,
    'anonymousDataCondition',
    { expression: Fn.conditionEquals('No', 'Yes') });

  // WHEN
  const solutionHelper = new SolutionHelper(stack, 'TestSolutionHelper', {
    sourceCodeBucketName: 'test-source-code-bucket',
    sourceCodeKeyPrefix: 'test-source-code-key-prefix',
    sendAnonymousData: 'Yes',
    anonymousUsageCondition: anonymousDataCondition,
    solutionVersion: 'beta',
    solutionId: 'TestSolution',
    deployM2CParameter: 'Yes'
  });

  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(solutionHelper.customResourceLambda).toBeDefined();
  expect(solutionHelper.anonymousDataUUID).toBeDefined();
  expect(solutionHelper.lowerCaseStackName).toBeDefined();
});
