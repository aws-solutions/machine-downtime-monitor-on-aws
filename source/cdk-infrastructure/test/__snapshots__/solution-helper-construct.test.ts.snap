// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Solution helper snapshot and variables 1`] = `
Object {
  "Conditions": Object {
    "anonymousDataCondition": Object {
      "Fn::Equals": Array [
        "No",
        "Yes",
      ],
    },
  },
  "Resources": Object {
    "TestSolutionHelperCustomResourceLambda6BB6EA27": Object {
      "DependsOn": Array [
        "TestSolutionHelperCustomResourceLambdaServiceRoleDC777AF2",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "CloudWatch permissions are granted by the LambdaBasicExecutionRole",
            },
            Object {
              "id": "W89",
              "reason": "VPC for Lambda is not needed. This serverless architecture does not deploy a VPC.",
            },
            Object {
              "id": "W92",
              "reason": "ReservedConcurrentExecutions is not needed for this Lambda function.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-source-code-bucket",
          "S3Key": "test-source-code-key-prefix/solution-helper.zip",
        },
        "Description": "Performs various Solution lifecycle actions when invoked as a CloudFormation Custom Resource",
        "Environment": Object {
          "Variables": Object {
            "ANONYMOUS_DATA_UUID": Object {
              "Fn::GetAtt": Array [
                "TestSolutionHelperGenerateSolutionConstantsCustomResource6D4DA57C",
                "AnonymousDataUUID",
              ],
            },
            "SEND_ANONYMOUS_DATA": "Yes",
            "SOLUTION_ID": "TestSolution",
            "SOLUTION_VERSION": "beta",
          },
        },
        "Handler": "solution-helper/index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperCustomResourceLambdaServiceRoleDC777AF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestSolutionHelperCustomResourceLambdaServiceRoleDC777AF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestSolutionHelperGenerateSolutionConstantsCustomResource6D4DA57C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Action": "GENERATE_SOLUTION_CONSTANTS",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperGenerateSolutionConstantsLambda5CE48C69",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestSolutionHelperGenerateSolutionConstantsLambda5CE48C69": Object {
      "DependsOn": Array [
        "TestSolutionHelperGenerateSolutionConstantsLambdaServiceRole9B3DCC97",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "CloudWatch permissions are granted by the LambdaBasicExecutionRole",
            },
            Object {
              "id": "W89",
              "reason": "VPC for Lambda is not needed. This serverless architecture does not deploy a VPC.",
            },
            Object {
              "id": "W92",
              "reason": "ReservedConcurrentExecutions is not needed for this Lambda function.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-source-code-bucket",
          "S3Key": "test-source-code-key-prefix/solution-helper.zip",
        },
        "Description": "Generates an anonymous UUID and a lower case stack name for sending metrics and resource creation",
        "Environment": Object {
          "Variables": Object {
            "STACK_NAME": Object {
              "Ref": "AWS::StackName",
            },
          },
        },
        "Handler": "solution-helper/index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperGenerateSolutionConstantsLambdaServiceRole9B3DCC97",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 3,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestSolutionHelperGenerateSolutionConstantsLambdaServiceRole9B3DCC97": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestSolutionHelperSolutionLifecycleMetricCustomResource14927B8D": Object {
      "Condition": "anonymousDataCondition",
      "Properties": Object {
        "Action": "SOLUTION_LIFECYCLE",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperCustomResourceLambda6BB6EA27",
            "Arn",
          ],
        },
        "SolutionParameters": Object {
          "DeployM2C": "Yes",
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
    },
  },
}
`;
